File: native/android/app/src/main/java/app/jim/guardian/ui/JimMiniOSMark.kt
package app.jim.guardian.ui

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * The jim-mini OS lockup — the whole mark, as the ABRACADABRA menu button.
 *
 * Three stacked parts in the proportions of the artwork: the neon wordmark,
 * the amulet triangle with ABRACADABRA descending eleven rows, and the OS
 * plate. The console draws the same lockup from the same numbers
 * (`app/src/JimMiniOS.tsx`, `assets/brand/jim-mini-os.svg`).
 *
 *     asked     is the mark on the button
 *     mattered  is the whole mark on the button, big enough to read as itself
 *
 * **Built from layout and one Path, not from a canvas with text drawn into
 * it.** Text on a canvas means reaching for `nativeCanvas` and measuring
 * glyphs by hand; `Text` in a `Column` is the same drawing with the layout
 * engine doing the arithmetic, and it is the shape this shell already knows
 * how to render. The only hand-drawn thing is the triangle, which is three
 * points.
 *
 * Every dimension is a fraction of the requested size, so the mark is one
 * mark whether it sits on a tab strip or fills a splash. The fractions are
 * passed down as a plain `Float` rather than read back off a layout: inside a
 * `Canvas` the name `size` means the draw scope's size, and a composable that
 * silently measured the wrong thing is the sort of defect that only shows up
 * on a device.
 */
private val ROWS: List<String> = (0 until 11).map { "ABRACADABRA".take(11 - it) }

private val NEON = Color(0xFFE9F04E)
private val PARCHMENT = Color(0xFFF4EFDC)
private val TRIANGLE_FILL = Color(0xFF15120A)
private val EDGE = Brush.linearGradient(
    listOf(Color(0xFFDFE94B), Color(0xFFF2C230), Color(0xFFB8861F)))

/** The share of the triangle's height the letters occupy. The bottom fifth
 *  is empty on purpose — on the amulet the descent stops above the apex. */
private const val LETTER_BAND = 0.78f

@Composable
fun JimMiniOSMark(size: Dp = 72.dp) {
    val px = size.value
    Box(
        Modifier.size(size).background(Color.Black),
        contentAlignment = Alignment.TopCenter,
    ) {
        Column(
            Modifier.fillMaxWidth().padding(vertical = (px * 0.05f).dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Top,
        ) {
            Text(
                "jim-mini",
                color = NEON,
                fontSize = (px * 0.15f).sp,
                fontStyle = FontStyle.Italic,
                fontFamily = FontFamily.Cursive,
                fontWeight = FontWeight.Medium,
                textAlign = TextAlign.Center,
                modifier = Modifier.height((px * 0.22f).dp),
            )
            AmuletTriangle(px)
            Box(
                Modifier
                    .padding(top = (px * 0.01f).dp)
                    .width((px * 0.82f).dp)
                    .height((px * 0.21f).dp)
                    .background(Color.White),
                contentAlignment = Alignment.Center,
            ) {
                Text("OS", color = Color.Black, fontSize = (px * 0.17f).sp,
                    fontWeight = FontWeight.Black)
            }
        }
    }
}

@Composable
private fun AmuletTriangle(px: Float) {
    val width = px * 0.42f
    val height = px * 0.37f
    Box(Modifier.width(width.dp).height(height.dp)) {
        Canvas(Modifier.matchParentSize()) {
            val path = Path().apply {
                moveTo(0f, 0f)
                lineTo(size.width, 0f)
                lineTo(size.width / 2f, size.height)
                close()
            }
            drawPath(path, TRIANGLE_FILL)
            drawPath(path, EDGE,
                style = Stroke(width = maxOf(1f, size.width * 0.05f)))
        }
        Column(
            Modifier
                .fillMaxWidth()
                .height((height * LETTER_BAND).dp)
                .padding(top = (height * 0.04f).dp),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            ROWS.forEach { row ->
                Text(
                    row,
                    color = PARCHMENT,
                    fontSize = (height * 0.075f).sp,
                    fontFamily = FontFamily.Serif,
                    maxLines = 1,
                    textAlign = TextAlign.Center,
                    modifier = Modifier.fillMaxWidth(),
                )
            }
        }
    }
}

File: native/ios/Sources/Views/JimMiniOSMark.swift
import SwiftUI

/// The jim-mini OS lockup — the whole mark, as the ABRACADABRA menu button.
///
/// Three stacked parts in the proportions of the artwork: the neon wordmark,
/// the amulet triangle with ABRACADABRA descending eleven rows, and the OS
/// plate. The console draws the same lockup from the same numbers
/// (`app/src/JimMiniOS.tsx`, `assets/brand/jim-mini-os.svg`).
///
///     asked     is the mark on the button
///     mattered  is the whole mark on the button, big enough to read as itself
///
/// **Built from stack views and one `Path`, not from a canvas with text drawn
/// into it.** Text on a canvas means measuring glyphs by hand on three
/// platforms that measure them differently; `Text` in a `VStack` is the same
/// drawing with the layout engine doing the arithmetic, and it is the shape
/// this shell already knows how to render. The only hand-drawn thing here is
/// the triangle, which is three points.
///
/// Every dimension is a fraction of `size`, so the mark is one mark whether it
/// is 44pt on a tab strip or 200pt on a splash.
struct JimMiniOSMark: View {
    var size: CGFloat = 72

    /// The eleven rows of the descent: ABRACADABRA, then one letter shorter
    /// each line, down to a single A.
    private static let rows: [String] = (0..<11).map {
        String("ABRACADABRA".prefix(11 - $0))
    }

    var body: some View {
        ZStack {
            Color.black
            VStack(spacing: 0) {
                Text("jim-mini")
                    .font(.system(size: size * 0.15, weight: .medium,
                                  design: .serif))
                    .italic()
                    .foregroundStyle(Color(red: 0.91, green: 0.94, blue: 0.31))
                    // Two shadows rather than one: a neon tube is a bright
                    // core inside a wider halo, and a single blur reads as a
                    // smudge instead.
                    .shadow(color: Color(red: 0.91, green: 0.94, blue: 0.31)
                        .opacity(0.9), radius: size * 0.02)
                    .shadow(color: Color(red: 0.85, green: 0.90, blue: 0.20)
                        .opacity(0.6), radius: size * 0.06)
                    .frame(height: size * 0.22)

                triangle
                    .frame(width: size * 0.42, height: size * 0.37)

                Rectangle()
                    .fill(.white)
                    .frame(width: size * 0.82, height: size * 0.21)
                    .overlay(
                        Text("OS")
                            .font(.system(size: size * 0.17, weight: .black))
                            .foregroundStyle(.black)
                    )
                    .padding(.top, size * 0.01)
            }
            .padding(.vertical, size * 0.05)
        }
        .frame(width: size, height: size)
        .accessibilityLabel("ABRACADABRA")
    }

    private var triangle: some View {
        GeometryReader { geo in
            let w = geo.size.width, h = geo.size.height
            ZStack {
                Path { p in
                    p.move(to: CGPoint(x: 0, y: 0))
                    p.addLine(to: CGPoint(x: w, y: 0))
                    p.addLine(to: CGPoint(x: w / 2, y: h))
                    p.closeSubpath()
                }
                .fill(Color(red: 0.08, green: 0.07, blue: 0.04))
                .overlay(
                    Path { p in
                        p.move(to: CGPoint(x: 0, y: 0))
                        p.addLine(to: CGPoint(x: w, y: 0))
                        p.addLine(to: CGPoint(x: w / 2, y: h))
                        p.closeSubpath()
                    }
                    .stroke(
                        LinearGradient(
                            colors: [Color(red: 0.87, green: 0.91, blue: 0.29),
                                     Color(red: 0.95, green: 0.76, blue: 0.19),
                                     Color(red: 0.72, green: 0.53, blue: 0.12)],
                            startPoint: .topLeading, endPoint: .bottomTrailing),
                        lineWidth: max(1, w * 0.05))
                )
                // The letters stop above the apex, as they do on the amulet —
                // the bottom fifth of the triangle is empty on purpose.
                VStack(spacing: 0) {
                    ForEach(Self.rows, id: \.self) { row in
                        Text(row)
                            .font(.system(size: h * 0.075, design: .serif))
                            .foregroundStyle(Color(red: 0.96, green: 0.94,
                                                   blue: 0.86))
                            .lineLimit(1)
                            .minimumScaleFactor(0.4)
                            .frame(maxWidth: .infinity)
                    }
                }
                .padding(.horizontal, w * 0.08)
                .frame(height: h * 0.78, alignment: .top)
                .frame(maxHeight: .infinity, alignment: .top)
                .padding(.top, h * 0.04)
            }
        }
    }
}
